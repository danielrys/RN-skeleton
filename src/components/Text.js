// @flow
import * as React from "react"
import { Text as RNText } from "react-native"

// types
import type { TextType } from "../types"

// utils
import renderStyle from "../utils/renderStyle"

// theme
import { Colors, Metrics } from "../themes"

const styles = {
  text: ({ color, lineHeight, size, centered }) => {
    const textColor = color ? { color } : {}
    const textSize = size ? { fontSize: size } : {}
    const textCenter = centered ? { textAlign: "center" } : {}
    const textLineHeight = lineHeight ? { lineHeight } : {}
    return {
      ...textColor,
      ...textSize,
      ...textCenter,
      ...textLineHeight,
    }
  },
}

type TextProps = {
  children: string | React.Node,
  centered?: boolean,
  color?: string,
  size?: number,
  lineHeight?: number,
  style?: TextType,
}

export default class Text extends React.PureComponent<TextProps> {
  static defaultProps = {
    centered: false,
    color: Colors.base,
    lineHeight: undefined,
    size: Metrics.texts.sizes.base,
    style: {},
  }

  render() {
    const { children, centered, color, size, lineHeight, style } = this.props
    return (
      <RNText
        style={[
          renderStyle(styles.text, {
            color,
            centered,
            lineHeight,
            size,
          }),
          style,
        ]}
      >
        {children}
      </RNText>
    )
  }
}
