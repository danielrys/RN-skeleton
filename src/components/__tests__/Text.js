import React from "react"
import renderer from "react-test-renderer"

import Text from "../Text"

describe("Text component", () => {
  it("renders with default props", () => {
    const tree = renderer.create(<Text>Default text</Text>).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("renders as bold", () => {
    const tree = renderer.create(<Text bold>Default bold text</Text>).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("renders with specified color and size", () => {
    const tree = renderer
      .create(
        <Text color="yellow" size={20}>
          Text with defined size and color
        </Text>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("renders with custom styles", () => {
    const tree = renderer
      .create(
        <Text style={{ color: "red", fontSize: 12 }}>
          Text with custom style
        </Text>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
