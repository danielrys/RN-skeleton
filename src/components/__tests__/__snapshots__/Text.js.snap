// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text component renders as bold 1`] = `
<Text
  style={
    Array [
      Object {
        "fontSize": 16,
      },
      Object {},
    ]
  }
>
  Default bold text
</Text>
`;

exports[`Text component renders with custom styles 1`] = `
<Text
  style={
    Array [
      Object {
        "fontSize": 16,
      },
      Object {
        "color": "red",
        "fontSize": 12,
      },
    ]
  }
>
  Text with custom style
</Text>
`;

exports[`Text component renders with default props 1`] = `
<Text
  style={
    Array [
      Object {
        "fontSize": 16,
      },
      Object {},
    ]
  }
>
  Default text
</Text>
`;

exports[`Text component renders with specified color and size 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "yellow",
        "fontSize": 20,
      },
      Object {},
    ]
  }
>
  Text with defined size and color
</Text>
`;
